src/ripple/overlay/impl/PeerImp.cpp

```diff
@@ -423,7 +424,10 @@ PeerImp::close()
 void
 PeerImp::fail(std::string const& reason)
 {
-    assert(strand_.running_in_this_thread());
+    if(! strand_.running_in_this_thread())
+        return strand_.post(std::bind (
+            (void(Peer::*)(std::string const&))&PeerImp::fail,
+                shared_from_this(), reason));
```

Comment:
My ignorance of the inner workings of `asio` is getting in my way here.  I'm hoping that we're very confident that the call to `strand_.post` makes a copy of `reason`.